### ISSUE TITLE
TDD week

### ISSUE THREAD

**Issue by MyPitit** [foundersandcoders/workshop-QMUL#11]
# Monday

**10:00 - 11:00** Introduction & Course Overview
**11:00 - 13:00** Github workshop
**13:00 - 14:00 Lunch**
**14:00 - 16:30** Introduction to HTML, CSS & GitHub Pages
**16:30 - 17:00** Upload to GH Pages
# Tuesday

**10:00 - 11:00** Introduction to TDD
Romanizer?
# Wednesday
# Thursday
# Friday

Project - Stopwatch build TDD
## Resources
- [Romanizer](https://docs.google.com/presentation/d/12c7YChUdiBKAi2kI8GVPA1XXCNQXkt4u-5O9BmnfqBY/edit#slide=id.gead6860a7_1_15)
- [Learn TDD](https://github.com/dwyl/learn-tdd)

**Comment by sohilpandya** [comment]
@MyPitit Thanks for creating this issue. 

@sofer @tormod17 @JackTierney 
Monday looks good. 

Tuesday we will first need to introduce Javascript to them, no need to do the Git/GitHub Repo as they need to focus of code craftmanship & TDD. 

# Tuesday

**10am - 1pm** Introduction to Javascript and the Basic Codewars Challenges without any tests just like we did with API Week. 
**2pm - 5pm** Introduction to TDD. What is TDD Why TDD and How TDD. I will create slides for this using the following repo - https://github.com/dwyl/learn-tdd
- This will include the intro to the Testing Suite we will be using for our tests. I am open to suggestions Jasmine or QUnit. I don't mind either, they are both very similar. 
- I will then give them an example of a function where we can do TDD and why it would be useful. i.e. `function add()` I will create it without any tests then realise it's actually error prone so will create tests to make it reliable hence showing them the red/green/refactor cycle in process. 
- Lastly to end the day we can give them another task, where they can create the function in TDD in pairs, strict Ping Pong style, test on test off. 

# Wednesday

**10am - 10.30am** Recap of TDD and why we should do it, talk about tech companies and the ethos they have in terms of using TDD. Good for them and their CV's. 
**10.30 - 1pm** Morning Challenge that Jack did with FAC8 and Ruth with FAC 7 very similar but will get them understand the testing suite and testing better. - https://github.com/JMurphyWeb/testing-morning-challenge
**2pm - 5pm** Coin Calculator Challenge from https://github.com/dwyl/learn-tdd

# Thursday

**10am -

**Comment by tormod17** [comment]
My suggestion is that we can introduce testing with the Kata's we create tests for some simple katas. Choose a testing library Qunit /  jasmine and write tests before we start the kata.

**Comment by MyPitit** [comment]
I can do a presentation about what is TDD, why, etc... and explaining how to set up Jasmine. 

How is that sounds @sohilpandya @tormod17 @JackTierney?

**Comment by JackTierney** [comment]
@sohilpandya doing UDACITY and using the content for Thursday / Wednesday. 
@jacktierny Codecademy
@tormod17 StopWatch and make it a Readme. also make coin calculator into a readme.

**Comment by iteles** [comment]
@MyPitit @JackTierney @sohilpandya @tormod17 Really worth someone trying to be at the stop go continue tomorrow and seeing what the feedback is from FAC8's TDD week to bring into QMUL TDD week!

**Comment by sohilpandya** [comment]
@iteles great shout! will make sure a couple of us are there tomorrow.😄 

@JackTierney @MyPitit @tormod17 I've been looking at the Udacity course and it's all Object orientated JS. Which I am not a fan of, thinking of doing Romanizer with the class instead of the TDD session from Udacity, lets chat tomorrow morning 💯

**Comment by MyPitit** [comment]
@iteles Sohil wanted to do a survey tomorrow morning to get some feedback :)

**Comment by MyPitit** [comment]
@iteles Sorry! I misunderstood your last comment. I thought it was for QMUL not for FAC8.

**Comment by MyPitit** [comment]
Katas:
https://repl.it/C1fG/1
https://repl.it/C1fF
https://repl.it/C1fH/1
https://repl.it/C2KO
https://repl.it/C2KO/3

**Comment by sohilpandya** [comment]
# Tuesday

**10am - 1pm** [JS talk](https://docs.google.com/presentation/d/1NNVIEdNtfRocjIZ1fQC6fsCmHrUXeQGZow6HldzEzT8/edit#slide=id.g147530f4fb_0_6) and challenges
**2pm - 3pm** [TDD Slide ](http://www.slideshare.net/MireiaSangalo/test-driven-development-63221704)
**3pm - 5pm** @JackTierney Codecademy.

# Wednesday

Need to convert from QUNIT to Jasmine.
**10am** Recap of TDD
**10.30am - 11am** [Jasmine Slide](http://www.slideshare.net/MireiaSangalo/jasmine-behaviourdriven-development)
**11am - 1pm** Simple TDD with Jasmine Challenges
  function add(a,b) with TDD
  2/3 more simple function with TDD
**2pm - 5pm** String calculator

# Thursday

**11-1pm** Jacks Morning Challenge from Qunit to Jasmine
**2pm - 5pm** Friday end. Stopwatch TDD Style. step by step instruction for it.

**Comment by sohilpandya** [comment]
String Calculator
Before you start
Try not to read ahead.
Do one task at a time. The trick is to learn to work incrementally.
Make sure you only test for correct inputs. There is no need to test for invalid inputs for this kata.
Method
Create a simple String calculator method Add(string numbers).
The method can take 0, 1 or 2 numbers, and will return their sum (for an empty string it will return 0) for example Add(“”) = 0 or Add(“1”) = 1 or Add(“1,2”) = 3
Start with the simplest test case of an empty string and move to 1 and 2 numbers.
Remember to solve things as simply as possible so that you force yourself to write tests you did not think about.
Remember to refactor after each passing test.
Allow the Add method to handle an unknown amount of numbers.
Allow the Add method to handle new lines between numbers (instead of commas).
The following input is ok: “1\n2,3” (will equal 6).
The following input is NOT ok: “1,\n” (not need to prove it - just clarifying).
Support different delimiters.
To change a delimiter, the beginning of the string will contain a separate line that looks like this: “//[delimiter]\n[numbers…]” for example “//;\n1;2” should return 3 where the default delimiter is ‘;’ .
Choosing a delimiter is optional. All existing scenarios should still be supported.
Calling Add with a negative number should return “negatives not allowed” + the negative that was passed. If there are multiple negatives, show all of them in the message.
Numbers bigger than 1000 should be ignored, so adding 2 + 1001 = 2.
Delimiters can be of any length with the following format: “//[delimiter]\n” for example: “//[???]\n1???2???3” should return 6.
Allow multiple delimiters like this: “//[delim1][delim2]\n” for example “//[£][%]\n1£2%3” should return 6.
Make sure you can also handle multiple delimiters with length longer than one character.
The Bowling Game Kata
The essence of this kata, popularized by Uncle Bob Martin, is to create a function bowlingScore that works out the score for 10-pin Bowling. The scoring system for bowling takes some getting used to, if you want to tackle this problem I would recommend looking up the rules on Wikipedia. The weird scoring system means this is a cool problem to tackle. Uncle Bob breaks this kata down into the following five tests:
Gutter game scores zero - When you roll all misses, you get a total score of zero.
All ones scores 20 - When you knock down one pin with each ball, your total score is 20.
A spare in the first frame, followed by three pins, followed by all misses scores 16.
A strike in the first frame, followed by three and then four pins, followed by all misses, scores 24.
A perfect game (12 strikes) scores 300.
However if you are feeling adventurous, you don’t have to take Uncle Bob’s test suggestions, you can just come up with your own tests.
I don’t have the answers as I didn’t do them myself, but you could ask FAC8 to show you their screencasts!
They were good problems for people starting out with TDD.
Especially the string one, the Bowling one was a lot harder.

**Comment by sohilpandya** [comment]
# String Calculator

### Before you start
- Try not to read ahead.
- Do one task at a time. The trick is to learn to work incrementally.
- Make sure you only test for correct inputs. There is no need to test for invalid inputs for this kata.
  
  ### Method
1. Create a simple String calculator method Add(string numbers).  
   - The method can take 0, 1 or 2 numbers, and will return their sum (for an empty string it will return 0) for example Add(“”) = 0 or Add(“1”) = 1 or Add(“1,2”) = 3
   - Start with the simplest test case of an empty string and move to 1 and 2 numbers.
   - Remember to solve things as simply as possible so that you force yourself to write tests you did not think about.
   - Remember to refactor after each passing test.
2. Allow the Add method to handle an unknown amount of numbers.
3. Allow the Add method to handle new lines between numbers (instead of commas).
   - The following input is ok: “1\n2,3” (will equal 6).
   - The following input is NOT ok: “1,\n” (not need to prove it - just clarifying).
   - Support different delimiters.
   - To change a delimiter, the beginning of the string will contain a separate line that looks like this: “//[delimiter]\n[numbers…]” for example “//;\n1;2” should return 3 where the default delimiter is ‘;’ .
   - Choosing a delimiter is optional. All existing scenarios should still be supported.
4. Calling Add with a negative number should return “negatives not allowed” + the negative that was passed. If there are multiple negatives, show all of them in the message.
5. Numbers bigger than 1000 should be ignored, so adding 2 + 1001 = 2.
6. Delimiters can be of any length with the following format: “//[delimiter]\n” for example: “//[???]\n1???2???3” should return 6.
7. Allow multiple delimiters like this: “//[delim1][delim2]\n” for example “//[£][%]\n1£2%3” should return 6.
8. Make sure you can also handle multiple delimiters with length longer than one character.  

# The Bowling Game Kata

The essence of this kata, popularized by Uncle Bob Martin, is to create a function bowlingScore that works out the score for 10-pin Bowling. The scoring system for bowling takes some getting used to, if you want to tackle this problem I would recommend looking up the rules on Wikipedia. The weird scoring system means this is a cool problem to tackle. Uncle Bob breaks this kata down into the following five tests:
1. Gutter game scores zero - When you roll all misses, you get a total score of zero.
2. All ones scores 20 - When you knock down one pin with each ball, your total score is 20.
3. A spare in the first frame, followed by three pins, followed by all misses scores 16.
4. A strike in the first frame, followed by three and then four pins, followed by all misses, scores 24.
5. A perfect game (12 strikes) scores 300.
However if you are feeling adventurous, you don’t have to take Uncle Bob’s test suggestions, you can just come up with your own tests.
